beluga/include/beluga/algorithm/spatial_hash.hpp

```diff
@@ -34,22 +34,29 @@ namespace beluga {
 
 namespace detail {
 
-/// Returns the floor of a value shifted.
+/// Returns the hashed and rotated floor of a value.
 /**
  * \tparam N Number of bits to be used from the integer result, the least significant will be used.
  * \tparam I Result will be shifted by I*N.
- * \param value Input value to be floored and shifted.
+ * \param value Input value to be hashed.
  * \return The calculated result.
  */
 template <std::size_t N, std::size_t I>
-constexpr std::size_t floor_and_shift(double value) {
-  // Compute the largest integer value not greater than value.
-  auto signed_value = static_cast<std::intmax_t>(std::floor(value));
-  // Compute the smallest unsigned integer equal to the source value modulo 2^n
-  // (where n is the number of bits used to represent the destination type).
-  auto unsigned_value = static_cast<std::uintmax_t>(signed_value);
-  // Create a fixed-size sequence of N bits and perform a binary shift left I * N positions.
-  return std::bitset<N>{unsigned_value}.to_ullong() << (I * N);
+constexpr std::size_t floor_and_fibo_hash(double value) {
+  constexpr auto kFib64 = 11400714819323198485LLU;  // golden ratio for 64 bits
```

Comment:
Can we assert that `std::size_t` is the same as `std::unit64_t`?
```cpp
static_assert(std::is_same_v<std::size_t, std::uint64_t>);
```

Reply:
Done!